pkg/flow/internal/controller/node_custom_component.go

```diff
@@ -0,0 +1,328 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/module"
+	"github.com/grafana/agent/pkg/flow/config"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+	"github.com/grafana/river/ast"
+	"github.com/grafana/river/vm"
+)
+
+// CustomComponentNode is a controller node which manages a custom component.
+//
+// CustomComponentNode manages the underlying custom component and caches its current
+// arguments and exports.
+type CustomComponentNode struct {
+	id                ComponentID
+	globalID          string
+	label             string
+	componentName     string
+	importLabel       string
+	declareLabel      string
```

Comment:
I find importLabel and declareLabel to be a little confusing; I understand that it's literally the label of the import block and the label of the declare block, but that's also one layer removed from the construct we're talking about.

I would rather this be named `importNamespace` and `customComponentName` which more accurately describes what they are doing from the perspective of CustomComponentNode.